### PR TITLE
Add a new sample application using Component

### PR DESCRIPTION
**Pull request by mtnygard**
See Stuart Sierra's [Component](https://github.com/stuartsierra/component) library for rationale and usage.

This sample also uses the new [Component.repl](https://github.com/stuartsierra/component.repl) add-on for dev workflow.

### CODE REVIEW

File: samples/vase-component/src/vase_component/main.clj [hunk 1]
```diff
@@ -0,0 +1,26 @@
+(ns vase-component.main
```
Comment (ddeaguiar) [inline samples/vase-component/src/vase_component/main.clj]: This ns is not referenced anywhere. Perhaps include `:main` in `project.clj`?

File: samples/vase-component/src/vase_component/main.clj [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  []
+  (if-let [s (System/getenv "PORT")]
+    (Long. s)
+    (throw (ex-info "Must set enviroment variable PORT" {}))))
+
+(defn -main
+  "Command-line main entry point."
+  []
+  (let [port   (env-port)
+        system (-> (system/system {})
```
Comment (ddeaguiar) [inline samples/vase-component/src/vase_component/main.clj]: `system/system` is zero-arity.

File: samples/vase-component/src/vase_component/main.clj [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+                   (update ::system/endpoint endpoint/with-port port)
```
Comment (ddeaguiar) [inline samples/vase-component/src/vase_component/main.clj]: `:endpoint`, not `::system/endpoint`.

File: samples/vase-component/src/vase_component/main.clj [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+                   component/start)]
+    (log/info :msg "System started" :port port)
+    (println "Service started at port" port)
+    (endpoint/join (::system/endpoint system))))
```
Comment (ddeaguiar) [inline samples/vase-component/src/vase_component/main.clj]: `:endpoint`, not `::system/endpoint`.

BTW, what's the benefit of `endpoint/join`?

File: samples/vase-component/src/vase_component/main.clj [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+(defn- env-port
+  "Gets port from environment variable, or throws."
+  []
+  (if-let [s (System/getenv "PORT")]
+    (Long. s)
+    (throw (ex-info "Must set enviroment variable PORT" {}))))
+
+(defn -main
+  "Command-line main entry point."
+  []
```
Comment (ddeaguiar) [inline samples/vase-component/src/vase_component/main.clj]: `[& args]` ?

**Comment by ddeaguiar** [comment]
Alternatively, just remove `vase-component.main`.

File: samples/vase-component/src/vase_component/main.clj (hunk 3 above)
Comment (mtnygard) [inline samples/vase-component/src/vase_component/main.clj]: Thanks.

File: samples/vase-component/src/vase_component/main.clj (hunk 2 above)
Comment (mtnygard) [inline samples/vase-component/src/vase_component/main.clj]: Thanks.

File: samples/vase-component/src/vase_component/main.clj (hunk 1 above)
Comment (mtnygard) [inline samples/vase-component/src/vase_component/main.clj]: Added to project.clj

File: samples/vase-component/src/vase_component/main.clj (hunk 5 above)
Comment (mtnygard) [inline samples/vase-component/src/vase_component/main.clj]: Added as `[& _]` because args are not used in the sample.

Comment (ohpauleez) [inline samples/vase-component/src/vase_component/main.clj]: Moved this to `[& args]` as that's the convention.  Those getting started with Vase (and Clojure/Pedestal/Datomic) may find `_` difficult to understand.

**Comment by ohpauleez** [comment]
This is good to merge, unless there are more additions you want to add @mtnygard

**Comment by mtnygard** [comment]
I think it's good to go.

**Comment by ddeaguiar** [comment]
+1

**Comment by ohpauleez** [comment]
Thanks all!

**Pull request merged**